topic/tests/index.md

```diff
@@ -508,4 +508,22 @@ Sometimes a directive may call an internal function that performs specific actio
 With the above example, we could combine the data sets with this example to provide even shorter code, but it may be undesirable to do so.  For instance, with this code structure one can assert additional effects happened or mix in several `beforeEach()` calls.  Remember that the goal is maintainable, clean and self-describing code.
 
 
+Functional Testing
+------------------
+
+Functional testing is the art of testing how your website or api works from the eye of a user, or at least with everything put together. It is very similar to unit testing in the way of how tests are laid out, but there are some things to watch out for.
+
+First, we want to load the page or refresh each time a test is started. This is to make sure we carry over any data from one test to another. For instance, if data was set to be used in the `session` we want to make sure it gets cleared out before loading the next test. This can usually be done in the `afterEach()` call using `window.sessionStorage.clear();`.
+
+Try to think what a user would do. Try running through the scenario you are setting up and pay attention to what you do. Forms can be a little touchy, as some people will click the button to submit the form, while others will tab to the submit button and hit enter, or just hit enter when in the last field. Try to write tests using different techniques to achieve the same goal, like submitting a form to make sure different ways work.
```

Comment:
A lot of these comments are for functional UI testing and don't apply as easily to functional API testing, though with some minor wording changes you could make them apply in both scenarios.  For instance, the paragraph above says "we want to load the page or refresh each time" and you could easily say "load the page, refresh a browser, start up the server or otherwise clear any data from memory".

As for this paragraph, "click the button to submit a form" and etc could be UI and it's fine to leave the UI-only example on here.  Maybe add something like "For functional tests without a UI, you would still try to imagine the types of situations that occur."